src/storage-types/src/sources/mysql.rs

```diff
@@ -278,15 +278,31 @@ pub struct MySqlSourceDetails {
         strategy = "proptest::collection::vec(any::<mz_mysql_util::MySqlTableDesc>(), 0..4)"
     )]
     pub tables: Vec<mz_mysql_util::MySqlTableDesc>,
-    /// The initial 'gtid_executed' set for the source. This is used as the effective
-    /// snapshot point, to ensure consistency if the source is interrupted but commits
-    /// one or more tables before the initial snapshot of all tables is complete.
+    /// The initial 'gtid_executed' set for each subsource. The index of each string in this
+    /// vector corresponds to the index of the corresponding table in the `tables` field.
+    /// If this vector has only one element, it is the initial gtid set for all tables.
+    ///
+    /// This is used as the effective snapshot point for each subsource to ensure correctness
+    /// if the source is interrupted but commits one or more tables before the initial snapshot
+    /// of all tables is complete.
     #[proptest(strategy = "any_gtidset()")]
-    pub initial_gtid_set: String,
+    pub initial_gtid_set: Vec<String>,
+}
+
+impl MySqlSourceDetails {
+    /// The initial 'gtid_executed' set for the given table index.
+    pub fn table_initial_gtid_set(&self, table_index: usize) -> &str {
+        self.initial_gtid_set.get(table_index).unwrap_or_else(|| {
+            self.initial_gtid_set
+                .first()
+                .expect("at least one initial gtid set")
+        })
```

Comment:
This has some risk of working when you ask for index 3 and we only have two. I would use the match over slices syntax to clearly express what's going on:
```suggestion
        match &*self.initial_gtid_set {
            [gtid] => gtid,
            gtids => gtids[table_index],
        }
```